service/graph.go

```diff
@@ -31,21 +31,202 @@ var _ pipelines = (*pipelinesGraph)(nil)
+			pipe.receivers[rcvrNode.ID()] = rcvrNode
+		}
+
+		pipe.capabilitiesNode = newCapabilitiesNode(pipelineID)
+
+		for _, procID := range pipelineCfg.Processors {
+			pipe.processors = append(pipe.processors, g.createProcessor(pipelineID, procID))
+		}
+
+		pipe.fanOutNode = newFanOutNode(pipelineID)
+
+		for _, exprID := range pipelineCfg.Exporters {
+			if set.Connectors.IsConfigured(exprID) {
+				connectorsAsExporter[exprID] = append(connectorsAsExporter[exprID], pipelineID)
+				continue
+			}
+			expNode := g.createExporter(pipelineID, exprID)
+			pipe.exporters[expNode.ID()] = expNode
+		}
+	}
+
+	for connID, exprPipelineIDs := range connectorsAsExporter {
+		for _, eID := range exprPipelineIDs {
+			for _, rID := range connectorsAsReceiver[connID] {
+				connNode := g.createConnector(eID, rID, connID)
+				g.pipelines[eID].exporters[connNode.ID()] = connNode
+				g.pipelines[rID].receivers[connNode.ID()] = connNode
+			}
+		}
+	}
+}
+
+func (g *pipelinesGraph) createReceiver(pipelineID, recvID component.ID) *receiverNode {
+	rcvrNode := newReceiverNode(pipelineID, recvID)
+	if node := g.componentGraph.Node(rcvrNode.ID()); node != nil {
+		return node.(*receiverNode)
+	}
+	g.componentGraph.AddNode(rcvrNode)
+	return rcvrNode
+}
+
+func (g *pipelinesGraph) createProcessor(pipelineID, procID component.ID) *processorNode {
+	procNode := newProcessorNode(pipelineID, procID)
+	g.componentGraph.AddNode(procNode)
+	return procNode
+}
+
+func (g *pipelinesGraph) createExporter(pipelineID, exprID component.ID) *exporterNode {
+	expNode := newExporterNode(pipelineID, exprID)
+	if node := g.componentGraph.Node(expNode.ID()); node != nil {
+		return node.(*exporterNode)
+	}
+	g.componentGraph.AddNode(expNode)
+	return expNode
+}
+
+func (g *pipelinesGraph) createConnector(exprPipelineID, rcvrPipelineID, connID component.ID) *connectorNode {
+	connNode := newConnectorNode(exprPipelineID.Type(), rcvrPipelineID.Type(), connID)
+	if node := g.componentGraph.Node(connNode.ID()); node != nil {
+		return node.(*connectorNode)
+	}
+	g.componentGraph.AddNode(connNode)
+	return connNode
+}
+
+func (g *pipelinesGraph) createEdges() {
+	for _, pg := range g.pipelines {
+		for _, receiver := range pg.receivers {
+			g.componentGraph.SetEdge(g.componentGraph.NewEdge(receiver, pg.capabilitiesNode))
+		}
+
+		var from, to graph.Node
+		from = pg.capabilitiesNode
+		for _, processor := range pg.processors {
+			to = processor
+			g.componentGraph.SetEdge(g.componentGraph.NewEdge(from, to))
+			from = processor
+		}
+		to = pg.fanOutNode
+		g.componentGraph.SetEdge(g.componentGraph.NewEdge(from, to))
+
+		for _, exporter := range pg.exporters {
+			g.componentGraph.SetEdge(g.componentGraph.NewEdge(pg.fanOutNode, exporter))
+		}
+	}
+}
+
+func (g *pipelinesGraph) buildComponents(ctx context.Context, set pipelinesSettings) error {
+	nodes, err := topo.Sort(g.componentGraph)
+	if err != nil {
+		return err // TODO clean up error message
```

Comment:
Can you be more specific about what is left here?

Reply:
When there is a cycle in the graph, there is enough information within the error to construct a better error message that indicates exactly the components that are in a cycle. It's non-trivial logic though, so I think it's best for another PR.